### ISSUE TITLE
SerperDevTool._run() missing 1 required positional argument: 'search_query'.

### ISSUE THREAD

**Issue by BracerJack** [crewAIInc/crewAI#430]
A.I says:
I encountered an error while trying to use the tool. This was the error: SerperDevTool._run() missing 1 required positional argument: 'search_query'.
 Tool Search the internet accepts these inputs: Search the internet(search_query: 'string') - A tool that can be used to semantic search a query from a txt's content.

I am just running from the example, copy and paste.
``
import os
from crewai import Agent, Task, Crew, Process
from crewai_tools import SerperDevTool

#os.environ["OPENAI_API_KEY"] = "YOUR_API_KEY"
os.environ["SERPER_API_KEY"] = "I have filled this in with a valid key" # serper.dev API key

os.environ["OPENAI_API_BASE"] ='http://localhost:11434/v1'
os.environ["OPENAI_MODEL_NAME"] ='mistral'  # Adjust based on available model
os.environ["OPENAI_API_KEY"] ='null' #switching between null and ollama makes no difference, the error as mentioned remains.

search_tool = SerperDevTool()
``

The rest is exactly copy and paste from the front page example.

**Comment by joaomdmoura** [comment]
Could be a model limitation but will give it a try as well!
Em 5 de abr. de 2024, 19:32 -0300, Bracer Jack ***@***.***> escreveu:

**Comment by BracerJack** [comment]
> Could be a model limitation

Ok let me try again now:
import os
from crewai import Agent, Task, Crew, Process
from crewai_tools import SerperDevTool

#os.environ["OPENAI_API_KEY"] = "YOUR_API_KEY"
os.environ["SERPER_API_KEY"] = "a0b7dc06666b68bac203ea16276b9e58c21e19dd" # serper.dev API key

# You can choose to use a local model through Ollama for example. See https://docs.crewai.com/how-to/LLM-Connections/ for more information.

os.environ["OPENAI_API_BASE"] ='http://localhost:11434/v1'
os.environ["OPENAI_MODEL_NAME"] ='llava' #I swtitch from 'mistral'  to llava.
os.environ["OPENAI_API_KEY"] ='null'
bla bla bla

Result:
Action 'search_query' don't exist, these are the only available Actions: Search the internet: Search the internet(search_query: 'string') - A tool that can be used to semantic search a query from a txt's content.

**Comment by joaomdmoura** [comment]
I'll also try the same models locally just in case to make sure we make it
better for them, but just in case, gpt4 works like a cham I guess?

--
João Moura
@joaomdmoura <http://twitter.com/joaomdmoura>


Em sex., 5 de abr. de 2024 às 19:38, Bracer Jack ***@***.***>
escreveu:

**Comment by BracerJack** [comment]
No idea, I only use local models.
CrewA.I. worked before, but I deleted the archived venv isolated folder to update because it has been a long time and that venv backup was old, should have kept the backup! ha ha ha
Good luck solving this :D

**Comment by BracerJack** [comment]
Switched from Ollama to LM Studio to see if anything changes, nope same issue:
Action '[Search the internet] 'artificial intelligence trends 2024' - A tool that can be used to semantic search a query from a txt's content.' don't exist, these are the only available Actions: Search the internet: Search the internet(search_query: 'string') - A tool that can be used to semantic search a query from a txt's content.

**Comment by BracerJack** [comment]
I have updated crewA.I. to the latest v0.28.1 7 April 2024 and it is still giving this error:
"I encountered an error while trying to use the tool. This was the error: SerperDevTool._run() missing 1 required positional argument: 'search_query'.
 Tool Search the internet accepts these inputs: Search the internet(search_query: 'string') - A tool that can be used to semantic search a query from a txt's content."

**Comment by NSMyself** [comment]
I've ran into the same issue but weirdly enough it only appears to fail sporadically; i first saw the error 5m ago and I ended up re-running just for the sake of it but it appears to be working now.

**Comment by NSMyself** [comment]
@joaomdmoura thanks for the quick reply. I've edited my comment above btw, it now appears to fail sporadically instead of permanently.

For the record, I'm using Mistral 7B on a MacBook Air M1 (16GB RAM). I'm aware it is a bit heavy, as there is some delay when chatting with the LLM locally. I will give it a shot on a small LLM later today hopefully.

**Comment by joaomdmoura** [comment]
Yup, smaller model will probably fail even more often

**Comment by NSMyself** [comment]
Oh, i see. Well, I'll try a larger one then 😅

**Comment by BracerJack** [comment]
Looks like the solution might be to have a list of "compatible models" along with the minimum quantization needed.

**Comment by 0xC0FFEE71** [comment]
I have the same issue here, SerperDevTool is not working, returning forementioned error message. Also here, I just copied and pasted the trip planner example, which worked well with gpt4. Now I am using Ollama on Windows and tried the following models:
llama2:7b (as far as I can see, the agent doesnt even try to use SerperDevTool, stops after max iterations)
llama2:13b (the agent tries to search, but SerperDevTool does not work)
llama2:70b (freezes, I can actually run models that large on my machine, but for this example, it doesn't work)
openhermes (SerperDevTool works sporadically, but then WebsiteSearchTool fails)
mistral:7b (works, SerperDevTool fails its very first attempt but then "adjusts" and runs well)

hope this helps

Do I still have to configure all model files to introduce "Result" as stop word?

**Comment by BracerJack** [comment]
I just ran it on dolphin-2.8-mistral-7b-v02, a recent model and it failed, but seeing from the comments that the failure/success rate is also "random", I cannot longer say it definitely doesn't work.    

The issue here is...it is not that the search is returning nothing...it is not that too little or too much information came from the search, the situation seem to be that the connection itself is not even established, you have:

1: "A tool that can be used to semantic search a query from a txt's content.' don't exist"
2: SerperDevTool._run() missing 1 required positional argument: 'search_query'.
3: Action 'search_query' don't exist, these are the only available~

The first one stats the tool didn't exist even though pip install "crewai[tools]" is used.
The second is a code error itself where 1 argument is missing, this seems to indicate code issue, not A.I. issue?
The third one, the function call doesn't exist...code issue again? Nothing to do with the A.I. ?

Any idea what is the version where this work? Or we have a case where even if we go back in versions it wouldn't help because the tools that are installed via pip install "crewai[tools]" wasn't archived so the old version of crew_ai would still be using the newer damaged tools?

I am speculating here, I have no idea.

**Comment by ThinksFast** [comment]
I'm also getting this error, but using the `JSONSearchTool`.   I have not tried this with other tools, but I've tried using both local LLMs and `gpt-4-0125-preview`.  

Full CrewAI error text is:
```
Action: Search a JSON's content

Action Input: {"query": "datadog"}

I encountered an error while trying to use the tool. This was the error: JSONSearchTool._run() missing 1 required positional argument: 'search_query'.
 Tool Search a JSON's content accepts these inputs: A tool that can be used to semantic search a query the data/knowledge_base.json JSON's content.
```

Environment:
```
python==3.11.8
crewai==0.28.6
crewai-tools==0.1.6
````

**Comment by thiagosrib** [comment]
Hi guys,

I have the same issue here using `codellama` local model with `ollama`.

`Action Input: {'string': 'current trends and breakthroughs in AI research'} 

I encountered an error while trying to use the tool. This was the error: SerperDevTool._run() missing 1 required positional argument: 'search_query'.
 Tool Search the internet accepts these inputs: Search the internet(search_query: 'string') - A tool that can be used to semantic search a query from a txt's content.`
 
 I'm trying to run the example on Readme file on the `creawAI github`.

**Comment by BracerJack** [comment]
Updated to the latest v0.28.5 as of 11 April 2024, same issue:
I encountered an error while trying to use the tool. This was the error: SerperDevTool._run() missing 1 required positional argument: 'search_query'.
 Tool Search the internet accepts these inputs: Search the internet(search_query: 'string') - A tool that can be used to semantic search a query from a txt's content.

**Comment by joaomdmoura** [comment]
> Do I still have to configure all model files to introduce "Result" as stop word?
@0xC0FFEE71  last time I check this was not necessary anymore


@BracerJack 
>Updated to the latest v0.28.5 as of 11 April 2024, same issue:
I encountered an error while trying to use the tool. This was the error: SerperDevTool._run() missing 1 required positional argument: 'search_query'.
Tool Search the internet accepts these inputs: Search the internet(search_query: 'string') - A tool that can be used to semantic search a query from a txt's content.

This one could be a bug we shipped a couple thins in the tools that I think impacted, this will get the team looking into it first thing in the morning

**Comment by BracerJack** [comment]
Thank you for taking the time to reply, it is nice to know our issues are heard.

**Comment by gcaith2** [comment]
Traceback (most recent call last):
  File "c:/Users/AVITA/Desktop/Learn Django/aiagent/newsletter.py", line 3, in <module>  
    from crewai_tools import SerperDevTool
ModuleNotFoundError: No module named 'crewai_tools'

**Comment by christianazinn** [comment]
Going to throw my hat into the ring and say it's affected me too - serving via LM Studio v0.2.19 beta, tested with Zephyr Beta 7B and Mistral Instruct 7B, failed both times. It doesn't even work with Command-R locally.

**Comment by sahilkamath0108** [comment]
I seem to have solved the problem. When you define the task that involves serper search tool, in the task description add this particular line: 'Set the input parameter as : search_query'. By default it sets the parameter as 'query' which is not allowed by serper, so.... try this

**Comment by theCyberTech** [comment]
Could you post a snippet example for reference?

**Comment by sahilkamath0108** [comment]
research_agent = Agent(
  role='Researcher',
  goal='Find and summarize the latest AI news. Set the input parameter as : search_query',
  backstory="""You're a researcher at a large company.
  You're responsible for analyzing data and providing insights
  to the business.""",
  verbose=True,
  llm=llm,
  tools=[search_tool]
)

task = Task(
  description='Find and summarize the latest AI news. Set the input parameter as : search_query',
  expected_output='A bullet list summary of the top 5 most important AI news',
  agent=research_agent
)

**Comment by sahilkamath0108** [comment]
research_agent = Agent(
  role='Researcher',
  goal='Find and summarize the latest AI news',
  backstory="""You're a researcher at a large company.
  You're responsible for analyzing data and providing insights
  to the business.""",
  verbose=True,
  llm=llm,
  tools=[search_tool]
)

task = Task(
  description='Find and summarize the latest AI news. Set the input parameter as : search_query',
  expected_output='A bullet list summary of the top 5 most important AI news',
  agent=research_agent
)

**Comment by NSMyself** [comment]
Adding `Set the input parameter as : search_query'` didn't really help in my case, even when using your agent and task:

 ```
[DEBUG]: == Working Agent: Researcher
 [INFO]: == Starting Task: Find and summarize the latest AI news. Set the input parameter as : search_query


> Entering new CrewAgentExecutor chain...
 Action: Search the internet(search_query="latest AI news")
Action Input: {'search_query': 'latest AI news'}

Action 'Search the internet(search_query="latest AI news")' don't exist, these are the only available Actions: Search the internet: Search the internet(search_query: 'string') - A tool that can be used to semantic search a query from a txt's content.

```

**Comment by carmineapice** [comment]
> I seem to have solved the problem. When you define the task that involves serper search tool, in the task description add this particular line: 'Set the input parameter as : search_query'. By default it sets the parameter as 'query' which is not allowed by serper, so.... try this

It worked apparently

**Comment by ananthpn** [comment]
> research_agent = Agent( role='Researcher', goal='Find and summarize the latest AI news', backstory="""You're a researcher at a large company. You're responsible for analyzing data and providing insights to the business.""", verbose=True, llm=llm, tools=[search_tool] )
> 
> task = Task( description='Find and summarize the latest AI news. Set the input parameter as : search_query', expected_output='A bullet list summary of the top 5 most important AI news', agent=research_agent )

Your solution worked for me, thanks much!

**Comment by BracerJack** [comment]
16 April 2024:
Adding "Set the input parameter as : search_query." doesn't work.

Now you get:
Action 'search_query' don't exist, these are the only available Actions: Search the internet: Search the internet(search_query: 'string') - A tool that can be used to semantic search a query from a txt's content.

**Comment by theCyberTech** [comment]
Quick question where are you setting the tool? On the agent or on the task?

**Comment by BracerJack** [comment]
I placed it in task following ananthpn's example :-)

**Comment by theCyberTech** [comment]
Ok, could you do me a favor and set it on the both the agent and task, it's the same attribute on the agent

**Comment by BracerJack** [comment]
I removed all traces of crewAI from my system, maybe at a later time :-)

**Comment by BracerJack** [comment]
> Ok, could you do me a favor and set it on the both the agent and task, it's the same attribute on the agent

It works. It works. It works!!!
I went full insane and paste:
Set the input parameter as : search_query.

Everywhere, in agent and task.
Now it works.
Alright that's it, I am making a backup of ALL these as it is working right now: 16 April 2024.

Thank you theCyberTech for your suggestion.

**Comment by theCyberTech** [comment]
No worries, FYI this also works for the rag tools, setting on the agent and the task

**Comment by BracerJack** [comment]
> FYI I'm a mod from the discord server so catch me on there if any other issues

I will if I ever get back to discord again, I was thinking of making a server for my YouTube subscribers but discord also take time and attention resources ;p

**Comment by vishalvihaapps** [comment]
It is not working still. I am using LLAMA2 7B via Ollama.

Error below:

Action 'Search the internet("Latest AI News")' don't exist, these are the only available Actions: Search the internet: Search the internet(search_query: 'string') - A tool that can be used to semantic search a query from a txt's content.

Request help on this issue.

Code is very simple:

`from crewai import Agent, Task, Crew, Process
from langchain_community.llms import Ollama
from langchain_community.tools import DuckDuckGoSearchRun
from langchain.tools import Tool
from crewai_tools import tool
from crewai_tools import SerperDevTool
from langchain_openai import ChatOpenAI

import os

os.environ["SERPER_API_KEY"] = "XXX"
os.environ["OPENAI_API_BASE"] = 'http://localhost:11434/v1'
os.environ["OPENAI_MODEL_NAME"] ='llama2' 
os.environ["OPENAI_API_KEY"] ='null'

search_tool = SerperDevTool()

research_agent = Agent(
  role='Researcher',
  goal='Find and summarize the latest AI news. Example: {"search_query": "Latest AI News"}',
  backstory="""You're a researcher at a large company.
  You're responsible for analyzing data and providing insights
  to the business.""",
  verbose=True,
  tools=[search_tool]
)

task = Task(
  description='Find and summarize the latest AI news. Example: {"search_query": "Latest AI News"}',
  expected_output='A bullet list summary of the top 5 most important AI news',
  agent=research_agent
)

crew = Crew(
    agents=[research_agent],
    tasks=[task],
    verbose=2
)

result = crew.kickoff()
print(result)`

**Comment by jimccadm** [comment]
Hi there.
This is 100% to do with Ollama/local LLMs.
I have confirmed this by testing llama3, llama2 (via Ollama) and the same search_query error persists.
On switching to OpenAI API key, 100% works fine, Serper does the search magic and zero errors.

**Comment by jimccadm** [comment]
UPDATE: After some further Ollama testing (mixtral, alfred, solar) I appear to be having success with dolphin2.1-mistral (4.1gb) with no need to add search_query addition(s). May be worth trying.

**Comment by elhakimz** [comment]
It sucessfully using phi3 (without changing temperature), but if I set temperature to .8 it failed to use SerperDevTool 
![image](https://github.com/joaomdmoura/crewAI/assets/616176/8d19315f-ca5e-44a6-ac43-f60fb37289ad)

**Comment by joaomdmoura** [comment]
Hey folks, good new this should be fixed on the new version we are now testing, `0.30.0rc3`, if you want ot give it a try you can do to with `pip install 'crewai[tools]'==0.30.0rc3` lmk how it goes!

**Comment by joaomdmoura** [comment]
It's non deterministic but we tried to make way easier for smaller models to use tools and this one in particular

**Comment by biraescudero** [comment]
João, thanks for your open source code and dedication.
I'm struggling with this problem. Using Groq and local models. 
I get an error installing this rc vesrion:
(.venv) C:\Users\ubira\PycharmProjects\teste_groq>pip install 'crewai[tools]'==0.30.0rc3
ERROR: Invalid requirement: "'crewai[tools]'==0.30.0rc3"

**Comment by joaomdmoura** [comment]
You can know also give the prompt templates that works the better for the model:

```python
Agent = Agent(
    role="{topic} specialist",
    goal="Figure {goal} out",
    backstory="I am the master of {role}",
    system_template="""<|start_header_id|>system<|end_header_id|>

{{ .System }}<|eot_id|>""",
    prompt_template="""<|start_header_id|>user<|end_header_id|>

{{ .Prompt }}<|eot_id|>""",
    response_template="""<|start_header_id|>assistant<|end_header_id|>

{{ .Response }}<|eot_id|>""",
)
```

**Comment by BracerJack** [comment]
3 May 2024
Testing 0.30.0rc3 as suggested.
Since error is supposedly fixed, I will now remove "Set the input parameter as : search_query." from the backstory parameter.

Error:
Action: Search the internet
Action Input: {'search_query': 'AI advancements 2024

{'message': 'Missing query parameter', 'statusCode': 400}

**Error Handling**

Since the `Search the internet` tool returned an error, I need to handle it before proceeding. Let me try rephrasing the search query:

Action: Search the internet
Action Input: {search_query: "latest advancements in AI 2024"}

Thought: Hopefully this revised query will yield better results Let's handle these errors and proceed with caution!

**Error Handling**

It seems that the `Action Input` is not being recognized as a valid key-value dictionary. Let me reformat it to match the expected input:
```
Action: Search the internet
Action Input: {'search_query': 'latest advancements in AI 2024'}
```
With this revised format, let's retry the search:
``Let me rephrase the Action Input one more time to ensure it matches the expected input:
```
Action: Search the internet
Action Input: {'search_query': 'latest advancements in AI 2024'}
```
I'll retry the search with this revised format...
I have stopped with the test, it doesn't work for me.

Fortunately I have made a venv backup of the version that does work with "Set the input parameter as : search_query." implemented so no loss there.

This is just the result of my own test, maybe others have better luck, I am giving this version a miss :-)

**Comment by biraescudero** [comment]
@BracerJack, try not pass query arguments and let the agent manage this.

My code:
search_tool = SerperDevTool(
    url = "https://google.serper.dev/search",
    country="br",
    location="location",
    locale="pt-br"
)

I hope it works fr you.

**Comment by BracerJack** [comment]
Thank you @biraescudero, the version I am using that works does not require me to fill in all these data in the SerperDevTool, I left it as search_tool = SerperDevTool() as the example shown for a yet an even older version that works when that example was created for.

For now I will stick to the old working version, only for now :D

**Comment by joaomdmoura** [comment]
@BracerJack this is super useful, what model are you using, I'd love to replicate it locally, but I think I found what the bug might be

**Comment by NSMyself** [comment]
@joaomdmoura firstly, thank you for being on top of this issue, it means a lot to me seeing you folks tackle these bugs so fast.

I am, however, still experiencing this issue when using Groq's  `llama3-70b-8192` and Serper:

```
Thought: I have gathered information about _some specific thing_

Action: None
Action Input: None

Action 'None' don't exist, these are the only available Actions: Search the internet: Search the internet(search_query: 'string') - A tool that can be used to search the internet with a search_query.

Thought: I need to analyze the search results to identify _some specific thing_

Action: None
Action Input: None

Action 'None' don't exist, these are the only available Actions: Search the internet: Search the internet(search_query: 'string') - A tool that can be used to search the internet with a search_query.

Thought: I need to analyze the search results to identify _some specific thing_

Action: None
Action Input: None

Action 'None' don't exist, these are the only available Actions: Search the internet: Search the internet(search_query: 'string') - A tool that can be used to search the internet with a search_query.

Thought: I need to analyze the search results to identify _some specific thing_

Action: None
Action Input: None

Action 'None' don't exist, these are the only available Actions: Search the internet: Search the internet(search_query: 'string') - A tool that can be used to search the internet with a search_query.

```

Occasionally, it just gets stuck like this. It happens when I ask to fetch some data and try to make sense of it. Just fetching data doesn't trigger it but requesting an opinion about the content of what it just fetched does.

I'm using the latest RC of crewai tools you posted above.

**Comment by NSMyself** [comment]
Quick update for my specific issue: it turns out that Groq was rate limiting my requests. I swapped to GPT4 and the error went away.

**Comment by mackimart1** [comment]
I created defined the search_query  for the serperDevTool and i'm not having the problem anymore. Hopefully it works for you guys too, tell me what you think and if any improves could be made i'm all ears  


_______________________________


import date
import time

def Search_the_internet(search_query, verbose=True):
    # Initialize the search tool
    search_tool = SerperDevTool()
    
    if verbose:
        print(f"Searching for: {search_query}")
    
    # Perform the actual search
    try:
        start_time = time.time()
        results = search_tool.search(search_query)
        end_time = time.time()
        
        if verbose:
            print(f"Search completed in {end_time - start_time:.2f} seconds.")
            print("Results:")
            for i, result in enumerate(results, 1):
                print(f"{i}. {result['title']}")
                print(f"   URL: {result['link']}")
                print(f"   Snippet: {result['snippet'][:100]}...")
                print()
        
        return results
    
    except Exception as e:
        if verbose:
            print(f"An error occurred during the search: {str(e)}")
        return None

**Comment by rendorHaevyn** [comment]
This was resolved for me (using the JSONSearchTool) by updating / adding latest CrewAI to my `requirements.txt`:

```txt
crewai==0.32.2
crewai[tools]
```

Pip freeze from previous fail state (search query error):
```txt
crewai==0.30.11
crewai-tools==0.2.6
```

Pip freeze from current success state:
```txt
crewai==0.32.2
crewai-tools==0.3.0
```

**Comment by rendorHaevyn** [comment]
This error is still occurring with JSONSearchTool, with the latest releases installed, without intervention.

```txt
crewai==0.32.2
crewai-tools==0.3.0
```

Per suggestions above, I've copy-pasta-ed the phrase "Set the input parameter as : search_query." into all `Task: description` and `Agent: goal` fields, and that appears to have addressed the issue.

**Comment by joaomdmoura** [comment]
I guess we could try to auto complete the wrongly formatted json a way to try to improve tool usage for smaller models, maybe that is a good compromise and work around for models struggling with formatting.
@rribeiro-ncode-pt did you tried this as well? https://github.com/joaomdmoura/crewAI/issues/430#issuecomment-2090888338

**Comment by joaomdmoura** [comment]
Maybe groq has a different default setting for things like temperature or such?

**Comment by sabre-code** [comment]
Hello, this is what worked for me. Added the following line in backstory of agents supposed to use serper tool

"When looking for information on internet use 'search_query' as input argument for search tools provided"
